src/main/java/works/crossdock/CrossdockServerInboundHandler.java

```
@@ -0,0 +1,135 @@
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import works.crossdock.client.Behavior;
+
+public class CrossdockServerInboundHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+  private final Map<String, Behavior> behaviors = new HashMap<String, Behavior>();
+
+  /** Populates all the different behaviors supported. */
+  public CrossdockServerInboundHandler(Map<String, Behavior> inputBehaviors) {
+    for (Entry<String, Behavior> entry : inputBehaviors.entrySet()) {
+      behaviors.put(entry.getKey(), entry.getValue());
+    }
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest fullHttpRequest)
+      throws Exception {
+    QueryStringDecoder queryStringDecoder = new QueryStringDecoder(fullHttpRequest.uri());
+    CrossdockRequest request = populateRequest(queryStringDecoder);
+    String behavior = request.getParam("behavior");
+    if (behavior == null) {
+      FullHttpResponse httpResponse =
+          new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+      ctx.writeAndFlush(httpResponse).addListener(ChannelFutureListener.CLOSE);
+      return;
+    }
+    Behavior br = behaviors.get(behavior);
+    runBehavior(br, request)
+        .whenComplete(
+            (response, ex) -> {
+              if (ex != null) {
+                ctx.fireExceptionCaught(ex);
+              } else {
+                writeAndCloseChannel(ctx, response);
+              }
+            });
+  }
+
+  /**
+   * Runs the request against the passed behavior.
+   *
+   * @param br behavior to run against
+   * @param req request to run
+   * @return CompletionStage that contains crossdockResponse
+   * @throws Exception if behavior throws an exception
+   */
+  public CompletionStage<CrossdockResponse> runBehavior(Behavior br, CrossdockRequest req)
+      throws Exception {
+    if (br == null) {
```

so Its skipped when the behavior doesn't exist? This is wrong. If the behavior doesn't exist, it should return a hard error. And this function runs a behavior, which by definition cannot be null. The caller for `runBehavior` is responsible for passing a valid behavior object.